src/cluster.h

```diff
@@ -173,257 +51,102 @@ typedef struct {
- * provide some information about the node state. */
-#define CLUSTERMSG_FLAG0_PAUSED (1<<0) /* Master paused for manual failover. */
-#define CLUSTERMSG_FLAG0_FORCEACK (1<<1) /* Give ACK to AUTH_REQUEST even if
-                                            master is up. */
-#define CLUSTERMSG_FLAG0_EXT_DATA (1<<2) /* Message contains extension data */
-
 /* ---------------------- API exported outside cluster.c -------------------- */
+/*
+ * Functions requiring per-clustering mechanism implementation: Lifecycle events
+ */
 void clusterInit(void);
 void clusterInitListeners(void);
 void clusterCron(void);
 void clusterBeforeSleep(void);
-clusterNode *getNodeByQuery(client *c, struct redisCommand *cmd, robj **argv, int argc, int *hashslot, int *ask);
-int verifyClusterNodeId(const char *name, int length);
-clusterNode *clusterLookupNode(const char *name, int length);
-int clusterRedirectBlockedClientIfNeeded(client *c);
-void clusterRedirectClient(client *c, clusterNode *n, int hashslot, int error_code);
-void migrateCloseTimedoutSockets(void);
 int verifyClusterConfigWithData(void);
+void clusterUpdateMyselfFlags(void);
 unsigned long getClusterConnectionsCount(void);
+void clusterUpdateMyselfHostname(void);
+void clusterUpdateMyselfAnnouncedPorts(void);
+void clusterUpdateMyselfIp(void);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: cluster commands
+ */
+void clusterCommand(client *c);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: Pub/Sub and module messages
+ */
 int clusterSendModuleMessageToTarget(const char *target, uint64_t module_id, uint8_t type, const char *payload, uint32_t len);
 void clusterPropagatePublish(robj *channel, robj *message, int sharded);
+void slotToChannelAdd(sds channel);
+void slotToChannelDel(sds channel);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: Debug and info actions
+ */
+void freeThisNodesLink(clusterNode *node);
+void freeNodeInboundLink(clusterNode *node);
+sds clusterGenNodesDescription(client *c, int filter, int use_pport);
+sds genClusterInfoString(void);
+
+/*
+ * Functions requiring per-clustering mechanism implementation: "Accessor" functions
+ */
+char* clusterNodeLastKnownIp(clusterNode *node);
+int clusterNodePort(clusterNode *node);
+clusterNode* clusterNodeGetSlaveof(clusterNode *node);
+int clusterNodeConfirmedReachable(clusterNode* node);
+int clusterNodeIsMaster(clusterNode* node);
+int clusterNodeIsSlave(clusterNode* node);
+int clusterNodeIsFailing(clusterNode* node);
+int clusterNodeTimedOut(clusterNode* node);
+int clusterNodeIsMyself(clusterNode* node);
+int clusterNodeIsNoFailover(clusterNode* node);
+clusterNode* getMyClusterNode(void);
+clusterNode* getNodeBySlot(int slot);
+clusterNode* getMigratingSlotDest(int slot);
+clusterNode* getImportingSlotSource(int slot);
+int isClusterHealthy(void);
+uint16_t getClusterNodeRedirectPort(clusterNode* node, int use_pport);
```

Comment:
Redirect isn't any different from normal port. If you look upstream you'll see we've refactored this mostly to be getClusterNodePlaintextPort and getClusterNodeTLSPort. That is the interface we want to expose to flotilla.

Reply:
So this does not quite allign with what you have upstream as it uses the use_pport param to decide which port to return. That is why I named the function after its usage to try and differentiate the name from just "give me this field." Open to new names but it's neither PlaintextPort or TLSPort.

Reply:
Let me try to add more color. The messaging the clusterbus sends around has a pport, because that is how it used to work so that is backwards compatible. The cluster node object has no pport, it just has TLS and plaintext. The decision we made is that we want the higher abstraction, the node, to just expose TLS and plaintext. See https://github.com/redis/redis/blob/unstable/src/cluster.c#L114.